src/server/routes/prefs.js

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/fluid-project/personal-data-server/blob/main/LICENSE
+ */
+
+"use strict";
+
+const express = require("express");
+const router = express.Router();
+const dbOps = require("../dbOps.js");
+
+/**
+ * Get preferences API
+ * Format: GET /get_prefs?loginToken=xxx
+ * 1. Verify if the loginToken is valid and not expired;
+ * 2. If it is, return the preferences;
+ * 3. If it isn't, return error.
+ */
+router.get("/get_prefs", async function (req, res) {
+    const loginToken = getLoginToken(req.headers.authorization);
+
+    if (!loginToken) {
+        res.status(403).json({"isError": true, "message": "Please login first."});
+        return;
+    }
+
+    const preferences = await dbOps.getPreferences(loginToken);
+
+    if (!preferences) {
+        res.status(403).json({"isError": true, "message": "Invalid login token. Please login."});
+        return;
+    } else {
+        res.json(preferences);
+    }
+});
+
+/**
+ * Save preferences API
+ * Format: POST /save_prefs
+ * Post data: the preferences object
+ * {
+ *     textSize: 1.2,
+ *     lineSpace: 1.3
+ *     ...
+ * }
+ * 1. Verify if the loginToken is valid and not expired;
+ * 2. Verify if the preferences object contains the right values;
+ * 3. If any verification failes, return error;
+ * 4. Otherwise, save preferences.
+ *
+ * Note: when the incoming body data is `undefined` or `null`, after going thru express.json(), req.body is
+ * assigned with an empty object. This empty object is then saved into the database.
+ * See: https://github.com/expressjs/body-parser/blob/master/lib/types/json.js#L74
+ */
+router.post("/save_prefs", async function (req, res) {
+    const loginToken = getLoginToken(req.headers.authorization);
+
+    if (!loginToken) {
+        res.status(403).json({"isError": true, "message": "Please login first."});
+        return;
+    }
+
+    // Verify if the login token is valid
+    const prevPrefs = await dbOps.getPreferences(loginToken);
+
+    if (!prevPrefs) {
+        res.status(403).json({"isError": true, "message": "Invalid login token. Please login."});
+        return;
+    }
+
+    const isSuccess = await dbOps.savePrefsByLoginToken(loginToken, req.body);
```

Comment:
Please add a comment at this line recording the discussion at the tech meeting - that is, that in the case that the client supplies an empty body, the JSON middleware will convert this into an empty object which will be saved, and link to the dodgy line https://github.com/expressjs/body-parser/blob/master/lib/types/json.js#L74 where this possibly surprising behaviour is implemented